src/internal/task/etcd_task_queue.go

```diff
@@ -0,0 +1,113 @@
+package task
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+
+	"github.com/cevaris/ordered_map"
+)
+
+const (
+	waitTime = 10 * time.Millisecond
+	bufSize  = 10
+)
+
+type taskFunc func()
+
+type groupEntry struct {
+	cancel       context.CancelFunc
+	taskFuncChan chan taskFunc
+}
+
+type taskQueue struct {
+	groups *ordered_map.OrderedMap
+	mu     sync.Mutex
+}
+
+func newTaskQueue(ctx context.Context) *taskQueue {
+	tq := &taskQueue{
+		groups: ordered_map.NewOrderedMap(),
+	}
+	go func() {
+		var cycleKey string
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			// Get the next group.
+			tq.mu.Lock()
+			iter := tq.groups.IterFunc()
+			kv, ok := iter()
+			tq.mu.Unlock()
+			var key string
+			if ok {
+				key = kv.Key.(string)
+			}
+			// Wait if there are no group entries or a cycle has been completed without finding a task.
+			if !ok || key == cycleKey {
+				time.Sleep(waitTime)
+				cycleKey = ""
+				continue
+			}
+			if cycleKey == "" {
+				cycleKey = key
+			}
+			// Check if the next task is ready for the group and process it.
+			ge := kv.Value.(*groupEntry)
+			select {
+			case cb := <-ge.taskFuncChan:
+				cb()
+				cycleKey = ""
+			default:
+			}
+			tq.requeueGroup(key)
+		}
+	}()
+	return tq
+}
+
+func (tq *taskQueue) group(ctx context.Context, groupID string, cb func(context.Context, chan taskFunc)) error {
```

Comment:
Is this method how you put tasks in the queue?

Reply:
This is how you set up a group in the internal task queue. Tasks are sent through the provided channel in the callback. This is used by the makers for managing the queuing of tasks across groups. Each group has a queue that is fed into this task queue which attempts to ensure fair execution of tasks across groups (tasks are the unit of scheduling, so it may not necessarily be fair in practice for groups with many small tasks vs groups with larger tasks, but I don't think that is a scheduling problem we should try to solve).